Improve isInputPending() docs and add info to pts api page (#27762)
mdn/content: files/en-us/web/api/scheduling/isinputpending/index.md

browser-compat: api.Scheduling.isInputPending
---

{{SeeCompatTable}}{{APIRef("Prioritized Task Scheduling API")}}

The **`isInputPending()`** method of the {{domxref("Scheduling")}} interface allows you to check whether there are pending input events in the event queue, indicating that the user is attempting to interact with the page.

This feature is useful in situations where you have a queue of tasks to run, and you want to yield to the main thread regularly to allow user interaction to occur so that the app is kept as responsive and performant as possible. `isInputPending()` allows you to yield only when there is input pending, rather than having to do it at arbitrary intervals.

## Syntax

```js-nolint
isInputPending()

> isInputPending(options)

Reviewer: Is this correct? You can't call this without the `navigator.scheduling` prefix.

Author: In the syntax box for each feature, we tend to just show the different possible configurations for the immediate feature and not the object chain that precedes it. It is clear from the code example how to use it, but I'll add a line to the top of the page just in case.

```

### Parameters

- `options` {{optional_inline}}
  - : An object providing options. Currently, the only option is:
    - `includeContinuous` {{optional_inline}}

>       - : A boolean, which is `false` by default. If set to `true`, it indicates that continuous events should be considered when checking for pending input. Continuous events are trusted events (which are events dispatched by the browser) that are fired successively, such as [`mousemove`](/en-US/docs/Web/API/Element/mousemove_event), [`wheel`](/en-US/docs/Web/API/Element/wheel_event), [`touchmove`](/en-US/docs/Web/API/Element/touchmove_event), [`drag`](/en-US/docs/Web/API/HTMLElement/drag_event), [`pointermove`](/en-US/docs/Web/API/Element/pointermove_event), and [`pointerrawupdate`](/en-US/docs/Web/API/Element/pointerrawupdate_event).

Reviewer: ```suggestion
      - : A {{jsxref("Boolean")}}, which is `false` by default. If set to `true`, it indicates that continuous events should be considered when checking for pending input. Continuous events are trusted events (which are events dispatched by the browser) that are fired successively, such as {{domxref("Element/mousemove_event", "mousemove")}}, {{domxref("Element/wheel_event", "wheel")}}, {{domxref("Element/touchmove_event", "touchmove")}}, {{domxref("HTMLElement/drag_event", "drag")}}, {{domxref("Element/pointermove_event", "pointermove")}}, and {{domxref("Element/pointerrawupdate_event", "pointerrawupdate")}}.
```

### Return value

> A boolean that indicates whether there are pending input events in the event queue (`true`) or not (`false`).

Reviewer: ```suggestion
A {{jsxref("Boolean")}} that indicates whether there are pending input events in the event queue (`true`) or not (`false`).
```

## Examples

We can use `isInputPending()` inside a task runner structure to run the `yield()` function only when the user is attempting to interact with the page:

```js
async function main() {
  // Create an array of functions to run
  const tasks = [a, b, c, d, e];

  while (tasks.length > 0) {
    // Yield to a pending user input
    if (navigator.scheduling.isInputPending()) {
      await yield();
    } else {
      // Shift the first task off the tasks array
      const task = tasks.shift();

      // Run the task
      task();
    }
  }
}
```

> This allows you to avoid blocking the main thread when the user is actively interacting with the page, potentially providing a smoother user experience.

Reviewer: ```suggestion
This allows you to avoid blocking the main thread when the user is actively interacting with the page, potentially providing a smoother user experience. However, by only yielding when necessary, we can continue running the current task when that is not necessary, therefore avoiding tasks being placed at the back of the queue behind any non-input related tasks that were scheduled after the current one.
```

## Browser compatibility

{{Compat}}

## See also

> - [Optimizing long tasks](https://web.dev/optimize-long-tasks/#yield-only-when-necessary) on web.dev (2022)

Reviewer: ```suggestion
- [Faster input events with Facebook’s first browser API contribution](https://engineering.fb.com/2019/04/22/developer-tools/isinputpending-api/) on engineering.fb.com (2019)
- [Better JS scheduling with isInputPending()](https://developer.chrome.com/articles/isinputpending/) on developer.chrome.com (2020)
- [Optimizing long tasks](https://web.dev/optimize-long-tasks/#yield-only-when-necessary) on web.dev (2022)
```
